### DISCUSSION TITLE
Upon using DataTable Directive in Angular, shows all data on first page itself (no pagination),on deleting any row it still persist on frontend upon page switch

### DISCUSSION THREAD

**Discussion by Harshad-Lukare** [l-lin/angular-datatables#1663] [Q&A]
Hi Team,

### Datatable pagination not working upon performing any activity

**1) ISSUE -1**

I am Using DataTable Directive in angular (version 13), So in this I fetch data from the pubic api "https://api.covidtracking.com/v1/us/daily.json" and all data shows properly with DataTable pagination 

But after refreshing page, it shows all data in first page itself  (IT SHOULD SHOW ALL DATA WITH PAGINATION)

All Data shown with Pagination:

![DataTable with pagination](https://user-images.githubusercontent.com/85783919/171839697-915d2afd-b59b-47c7-8321-3f8c0807d68f.png)

After refresh Data shown by DataTable on one Page only:

![Datatable not pagination-1](https://user-images.githubusercontent.com/85783919/171840046-ef5f6ac1-d03e-4ed5-8c05-80b39f03e753.png)

![datatable not pagination-2](https://user-images.githubusercontent.com/85783919/171840074-8cf3c86c-2447-4804-b17f-6801ea9c04aa.png)

**2) ISSUE 2:**

When we click on delete button, selected row gets filter from data array, row gets deleted from frontend also. But upon page switch and come back to page, deleted row still persist in frontend view.

![delete row in DT-1](https://user-images.githubusercontent.com/85783919/171840761-4ba2813b-446e-449f-a833-8b35d1f99ffc.png)

After switch to another page and then back to previous page

![DT row deletion-2](https://user-images.githubusercontent.com/85783919/171840962-28e5ede5-824d-4d45-885d-a6166ccfaa0e.png)

BELOW IS MY CODE:

File 1:html file

![file1](https://user-images.githubusercontent.com/85783919/171842030-3b589622-3534-484d-9e03-99363c2d32e4.png)

File 2:  .ts file
![file2](https://user-images.githubusercontent.com/85783919/171842062-2102ce1d-09cc-4756-bc54-0e82f0d25cd4.png)

![file2 Continued](https://user-images.githubusercontent.com/85783919/171842080-874617ec-6431-46fd-932e-c62a7e4da39d.png)

![file2 continued 3](https://user-images.githubusercontent.com/85783919/171842091-507d7a66-4f11-4648-a06a-e5807fc5cec8.png)

**Please check why DataTable shows this behavior upon deleting any row. Suggest some api by using which we solve pagination 
issue in angular(v-13)**

**Comment by shanmukhateja** [comment]
Hello,

Try using this example https://l-lin.github.io/angular-datatables/#/basic/server-side-angular-way and see if this is fixed for you.

Can you provide a reproducible StackBlitz repo to investigate this?

Also, please note, I'm moving this issue as a "discussion".

**Comment by shanmukhateja** [comment]
The best way to resolve this problem is to initialize the table AFTER you fetch the data from server.

1. Move the table init code inside ngOnInit to a separate function.
2. After line 52, once `this.posts` is updated, call the table init function you've created in the above step and let the table initialize.

Hope this helps :)

**Comment by mtholen** [comment]
You may have to rerender the table when new data comes in or when the page refreshes.
I do this usually in the subscription return, for example in my supplier list component:
```
ngOnInit(): void {
    this.initListeners();
    this.cs.getSupportTypes([SupportType.supplier]);
  }

  // cs = CommonService, is an angular service that controls the data fetching.
  initListeners() {
    this.cs.suppliers$.subscribe((s) => {
      this.suppliers = s;
      this.dtRender();
    });
  }
```
My `dtRender()` and `dtDestroy` look like this:

```
dtDestroy(): void {
    try {
      if (this.dtElement?.dtInstance) {
        this.dtElement.dtInstance.then((dtInstance: DataTables.Api) => dtInstance.destroy());
      }
    } catch (e) {
      console.error(e);
    }
  }

  dtRender(): void {
    try {
      this.dtDestroy();
      if (this.dtTrigger.observed) this.dtTrigger.next(null);
    } catch (e) {
      console.error(e);
    }
  }
```

So, in essence, I destroy and rerender the table on every new dataset that comes through.
